Client/Frontend/Settings/SearchCustomEngineViewController.swift

```diff
@@ -0,0 +1,727 @@
+    }
+    
+    private func checkManualAddExists() {
+        guard let url = urlText, let title = titleText  else {
+            return
+        }
+        
+        if !url.isEmpty, !title.isEmpty {
+            changeAddButton(for: .enabled)
+        } else {
+            changeAddButton(for: .disabled)
+        }
+    }
+}
+
+// MARK: - UITextViewDelegate
+
+extension SearchCustomEngineViewController: UITextViewDelegate {
+    
+    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        guard text.rangeOfCharacter(from: .newlines) == nil else {
+            textView.resignFirstResponder()
+            return false
+        }
+
+        return textView.text.count + (text.count - range.length) <= Constants.urlEntryMaxCharacterCount
+    }
+    
+    func textViewDidChange(_ textView: UITextView) {
+        changeAddButton(for: .disabled)
+        
+        urlText = textView.text
+
+        if let text = textView.text?.trimmingCharacters(in: .whitespacesAndNewlines),
+           let encodedText = text.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed),
+           let url = URL(string: encodedText),
+           url.host != nil,
+           url.isWebPage() {
+            if let scheme = url.scheme, let host = url.host {
+                self.host = URL(string: "\(scheme)://\(host)")
+            }
+        }
+    }
+
+    func textViewDidEndEditing(_ textView: UITextView) {
+        urlText = textView.text
+    }
+}
+
+// MARK: - UITextFieldDelegate
+
+extension SearchCustomEngineViewController: UITextFieldDelegate {
+    
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        guard let text = textField.text, text.rangeOfCharacter(from: .newlines) == nil else { return false }
+        
+        let currentString = text as NSString
+        let newString: NSString = currentString.replacingCharacters(in: range, with: string) as NSString
+        
+        let shouldChangeCharacters = newString.length <= Constants.titleEntryMaxCharacterCount
+        
+        if shouldChangeCharacters {
+            titleText = newString as String
+            checkManualAddExists()
+        }
+        
+        return shouldChangeCharacters
+    }
+    
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        titleText = textField.text
+    }
+
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        textField.endEditing(true)
+        
+        return true
+    }
+}
+
+// MARK: - SearchEngineTableViewHeader
+
+fileprivate class SearchEngineTableViewHeader: UITableViewHeaderFooterView {
+    
+    // MARK: Design
+    
+    struct Design {
+        static let headerHeight: CGFloat = 44
+        static let headerInset: CGFloat = 20
+    }
+    
+    // MARK: Properties
+    
+    var titleLabel = UILabel().then {
+        $0.font = UIFont.systemFont(ofSize: 14)
```

Comment:
If we can support dynamic type font here it would be good but would require updating how height is calculated since it looks like its constant right now

Reply:
Changes related with dynamic type font and  height calculation is made.